### PR TITLE
modules: Basic binary blob infrastructure

### PR DESCRIPTION
**Pull request by carlescufi**
This patch introduces the basic infrastructure to list and fetch binary
blobs. This includes:

- The new 'blobs' extension command
- An implementation of the `west blobs list` command with custom formatting
- A very simple mechanism for loading fetchers
- A basic implementation of an HTTP fetcher

Sample `module.yml`:

```
name: hal_nordic
build:
  cmake-ext: True
  kconfig-ext: True
blobs:
  - path: bloh.a
    sha256: ba148ca83988fca33d900a8584e3f667d575b7756732d667e02604909e1b726a
    type: lib
    version: '1.1'
    license-path: zephyr/blob-lic.txt
    url: https://raw.githubusercontent.com/zephyrproject-rtos/zephyr/main/boards/arm/96b_neonkey/doc/img/96b_neonkey.jpg
    description: foo
    doc-url: https://www.example.com
  - path: blah.bin
    sha256: 631d72c5fa12c5d5ab7d0f9c94a4e676d14250a85059615e29cd86755200feb1
    type: img
    version: '1.4'
    license-path: zephyr/blob-lic.txt
    url: https://raw.githubusercontent.com/zephyrproject-rtos/zephyr/main/boards/arm/96b_nitrogen/doc/img/96b_nitrogen.png
    description: bar
    doc-url: https://www.example.com
```

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
cc @catch-twenty-two

File: scripts/west_commands/blobs.py
Comment (mbolivar-nordic) [inline scripts/west_commands/blobs.py:36]: Use of colons, parentheses, and commas makes this default format string unfriendly to tools like cut and awk. Did you consider that and choose this anyway?

File: scripts/west_commands/blobs.py
Comment (mbolivar-nordic) [inline scripts/west_commands/blobs.py:74]: Do you plan on doing this?

File: scripts/west_commands/blobs.py
Comment (mbolivar-nordic) [inline scripts/west_commands/blobs.py:98]: Huh. I'm surprised so much boilerplate is required, but I can't find a simpler way in the hashlib documentation.

File: scripts/west_commands/blobs.py
Comment (mbolivar-nordic) [inline scripts/west_commands/blobs.py:81]: I'm happy to defer to your decision since you're going to maintain this, but did you consider using `default=[], action='append'` instead? That would let users do `west blobs list -m module1 -m module2` instead of `west blobs list -m module1,module2`, which I think is a bit nicer. It doesn't make any assumptions about module names not containing commas, and it makes it easier for scripts to add on modules by tacking on additional arguments without having to worry about string joining.

File: scripts/west_commands/fetchers/__init__.py
Comment (mbolivar-nordic) [inline scripts/west_commands/fetchers/__init__.py:1]: ?

File: scripts/west_commands/fetchers/fetcher.py
Comment (mbolivar-nordic) [inline scripts/west_commands/fetchers/fetcher.py:12]: Might want to include a request to preserve alphabetical order before this gets out of hand

File: scripts/west_commands/fetchers/http.py
Comment (mbolivar-nordic) [inline scripts/west_commands/fetchers/http.py:20]: Please clean this up

File: scripts/west_commands/fetchers/http.py
Comment (mbolivar-nordic) [inline scripts/west_commands/fetchers/http.py:5]: Just FYI, the usual import order in the west_commands directory is:

1. standard library
2. third party from pypi
3. third party from within zephyr

So this would instead be:

```py
from abc import ...

import requests
from west ...

from fetches.fetcher ...
```

Again I won't insist since you're going to maintain this, but consistency would be better here IMO

File: scripts/zephyr_module.py
Comment (mbolivar-nordic) [inline scripts/zephyr_module.py:121]: I guess this will become `required: false` when another fetcher is implemented?

**Review by mbolivar-nordic — COMMENTED** [review]
Overall this looks quite reasonable and in line with what the TSC voted to approve.

A few comments I noticed.

File: scripts/west_commands/blobs.py
Comment (ifyall) [inline scripts/west_commands/blobs.py:116]: Values like 'module' and 'status' are used for indexing into blob repeatedly. Should there be a file-scoped variable that contains these names rather than using them directly?

Comment (carlescufi) [reply scripts/west_commands/blobs.py:116]: We could, but I just didn't think it was worth it, given the simplicity of the code. I would start with this and then, if complexity grows, create "macros" for them.

**Comment by carlescufi** [comment]
> Overall this looks quite reasonable and in line with what the TSC voted to approve.

Thanks for the early review and apologies for pushing a half-baked version, this should've been a draft PR.

Note that once this PR goes in I will then submit another one documenting formally the command and the `module.yml` schema.

File: scripts/west_commands/blobs.py
Comment (carlescufi) [reply scripts/west_commands/blobs.py:34]: I did, because this tool has an `-f/--format` param which you can use to format the output in any way you want. However, I do agree that, for consistency with `west list`, plain space-separated items would be better. I will make the change.

File: scripts/west_commands/blobs.py
Comment (carlescufi) [reply scripts/west_commands/blobs.py:72]: Yes, but I will do a wholesale update of the completion script in a follow-up PR.

File: scripts/west_commands/blobs.py
Comment (carlescufi) [reply scripts/west_commands/blobs.py:79]: Very good suggestion, thanks! I will implement it.

File: scripts/west_commands/fetchers/__init__.py
Comment (carlescufi) [reply scripts/west_commands/fetchers/__init__.py]: sorry, fixed

File: scripts/west_commands/fetchers/http.py
Comment (carlescufi) [reply scripts/west_commands/fetchers/http.py]: fixed

File: scripts/west_commands/fetchers/http.py
Comment (carlescufi) [reply scripts/west_commands/fetchers/http.py:5]: Thank you! I will definitely fix this and remember it for the future.

File: scripts/zephyr_module.py
Comment (carlescufi) [reply scripts/zephyr_module.py:121]: No, all fetchers will use URLs. And the URL itself is required to identify the fetcher.

File: scripts/west_commands/fetchers/fetcher.py
Comment (carlescufi) [reply scripts/west_commands/fetchers/fetcher.py:12]: I have rewritten this bit instead

File: scripts/zephyr_module.py
Comment (tejlmand) [inline scripts/zephyr_module.py:132]: when you are using `.with_suffix('.yml')` you can actually keep the suffix here for readability (and indirectly provide a preference), as the `.with_suffix()` will substitute the suffix.

Like this:
```
MODULE_YML_PATH = PurePath('zephyr/module.yml')
```

and then later:
```
 for module_yml in [module_path / MODULE_YML_PATH,
                       module_path / MODULE_YML_PATH.with_suffix('.yaml')]:
```
which will substitute `.yml` with `.yaml`.

Ref: https://docs.python.org/3/library/pathlib.html
> Return a new path with the [suffix](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.suffix) changed. If the original path doesn’t have a suffix, the new suffix is appended instead. If the suffix is an empty string, the original suffix is removed:
> ```
> >> p = PureWindowsPath('c:/Downloads/pathlib.tar.gz')
> >> p.with_suffix('.bz2')
> PureWindowsPath('c:/Downloads/pathlib.tar.bz2')
> ```

File: scripts/zephyr_module.py
Comment (tejlmand) [inline scripts/zephyr_module.py:113]: we have two types, `img` and `lib`, but I don't see how those types are used or why they are there.

Do we need to differentiate and why ?
I assume that somewhere in the build system itself (CMake) some dedicated code for handling a particular binary blob will be written, and that specific code will know what to do with a given blob, regardless of the type specified in the yaml blob.

Or do you envision some generic handler that can automatically perform an action if the bin blob is of a dedicated type in the future.

**Review by tejlmand — COMMENTED** [review]
Thanks for taking up this task.

Looks good, just a couple of questions / comment, nothing blocking.

File: scripts/zephyr_module.py
Comment (carlescufi) [reply scripts/zephyr_module.py]: Very nice suggestion, implemented. Thanks!

File: scripts/zephyr_module.py
Comment (carlescufi) [reply scripts/zephyr_module.py:113]: Right now this is exclusively for the users' information. Every time a user types `west list` they will see the type(s) of blob each HAL contains, and then it's up to them to decide whether they want to use blobs or not.

File: scripts/west_commands/fetchers/__init__.py
Comment (carlescufi) [inline scripts/west_commands/fetchers/__init__.py:33]: Unsure why @mbolivar-nordic did not do the same in the runners module:
https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/west_commands/runners/__init__.py#L54

But I personally prefer this approach to the list of names that needs to be maintained.

File: scripts/zephyr_module.py
Comment (tejlmand) [reply scripts/zephyr_module.py:113]: thanks for confirming my observation.

I don't have a problem with the field, as it purely provides informational value, but curious if it then should be required or optional.
Of course using required from start, avoids everyone later having to update their blob yaml.
Allowing us to easier implement generic use of the field as all blobs already contains the field.
Having `enum` does provide uniform way of describing this field (avoiding differences like `lib` vs `library`).
On the other hand, for something being informational, then we need to update the `enum` every time it's not precise enough.

Consider someone adding both `32-bit` and `64-bit` version, they might like to have: `lib32` and `lib64` as available types.

So trying to find the right balance of not over-engineering from start for something we are not using today (and thus might need to adjust anyway when we start to use it), while at the same time being sufficient prepared for the future.

**Review by tejlmand — DISMISSED** [review]
Thanks for this, lgtm.

I see pros/cons by introducing `type` field, but will leave the final decision to the author.

File: scripts/zephyr_module.py
Comment (carlescufi) [reply scripts/zephyr_module.py:113]: I think `lib` and `img` are a good starting point. I doubt very much that we will need to differentiate between 32 and 64-bit in the actual type, but this can instead be achieved by adding a new `arch` field in the schema in the future, which I think makes more sense than using the `type` field.

**Comment by carlescufi** [comment]
@rettichschnidi FYI

File: scripts/zephyr_module.py
Comment (tejlmand) [reply scripts/zephyr_module.py:113]: > I doubt very much that we will need to differentiate between 32 and 64-bit in the actual type, but this can instead be achieved by adding a new arch field in the schema in the future

unfortunately my :crystal_ball: is broken.

File: scripts/west_commands/fetchers/__init__.py
Comment (mbolivar-nordic) [reply scripts/west_commands/fetchers/__init__.py:33]: I guess it's really a question of allowlist (the runners approach) and blocklist (the fetchers approach). Either one works and I guess it's up to taste to some extent.

Comment (carlescufi) [reply scripts/west_commands/fetchers/__init__.py:33]: Makes sense, thanks for the insight.

Comment (mbolivar-nordic) [reply scripts/west_commands/fetchers/__init__.py:33]: Did you really mean to make this a separate commit in the same PR? Why?

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Please squash the commits where you change your approach. If you want to leave a record of the reasoning, you can use comments or the commit log. Taking one approach, then immediately changing course, is not good branch structure IMO

File: scripts/west_commands/fetchers/__init__.py
Comment (carlescufi) [reply scripts/west_commands/fetchers/__init__.py:33]: Yes, I did this intentionally. I thought it was valuable to keep a record of my thought process when creating the infrastructure.

**Review by mbolivar-nordic — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]
lgtm, re-approved

**Review by ifyall — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
I think this is ready to go in given that it followed from several months of detailed discussions at TSC level. However, @henrikbrixandersen, since you self-requested review, I'm going to leave this open for a day or two more before merging. We can of course address any issues discovered after merge as well if you don't have cycles in that time.

File: scripts/west_commands/fetchers/__init__.py
Comment (henrikbrixandersen) [inline scripts/west_commands/fetchers/__init__.py:33]: This "suffers" from the same limitation as the runner discovery method - it only supports in-tree fetchers. While I do admit that out-of-tree fetchers are likely a more curious thing than out-of-tree runners, it would be nice if we could derive a better pattern for this.

**Review by henrikbrixandersen — APPROVED** [review]

File: scripts/west_commands/fetchers/__init__.py
Comment (carlescufi) [reply scripts/west_commands/fetchers/__init__.py:33]: It does, yes. But in this case it’s not an issue because an out of tree fetcher just doesn’t really make sense. You might as well commit your binaries to you OoT module.

**Pull request merged**
